All the things!

scull7 opened the pull request:
- update generic pool to version 3
- create an abstraction for the rabbit and couch pools
- log any errors coming from the pool initialization
- change all pool functions to use the new promise based API
- fix shutdown methods
- make sure to release resources properly
- add back off to pool initialization so that we don’t flood the
network with connection requests.
- add debugging
- all this and more!

scull7 committed:
All the things!

- update generic pool to version 3
- create an abstraction for the rabbit and couch pools
- log any errors coming from the pool initialization
- change all pool functions to use the new promise based API
- fix shutdown methods
- make sure to release resources properly
- add back off to pool initialization so that we don’t flood the
network with connection requests.
- add debugging
- all this and more!

scull7 commented:
@eostretsov This also fixes the "I only add new fields" issue within the CouchDB integration

jonajonlee reviewed (commented):
Does the couch config change at all in the defaults.yml?

scull7 commented:
No, this should be transparent to the outside 🌎

scull7 commented:
Except that the connection pool methods are all now promise based.  

pool.acquire and pool.release being the methods of interest

jonajonlee commented:
Sweet, looks good!

jonajonlee reviewed (approved):
:shipit:

jxbadam commented on lib/pool.js:
```diff
@@ -0,0 +1,93 @@
+
+const addBackoff = (factory) => {
+  const backoff   = new FibonacciBackoff()
+  const orgCreate = factory.create
```
orgCreate?

scull7 replied:
The original create method.  I'm following a pattern found in the express.js code

jxbadam reviewed (approved):
:shipit:

ricejasonf commented on lib/pool.js:
```diff
@@ -0,0 +1,93 @@
+}
+
+
+const buildFactory = (builder, default_config) => R.compose(
```
Why not auto-curry these functions that seem to be arbitrarily tupled/curried? With the way that javascript will allow this first function to take three params and silently discard the third leads to very annoying, difficult to track bugs.

mattsmee reviewed (approved):
:shipit:

scull7 committed:
improve stack traces

- change validator prop to ‘validate’ from ‘validateAsync’.  A look
through the ‘generic-pool’ code shows that ‘validateAsync’ is not a
valid option no matter what the docs say :)

scull7 committed:
Improve stack trace output

- change most anonymous functions to be named functions.
- remove arbitrary function currying
- mark functions as private using the dangling ‘_’ character.
- multiple ‘_’ indicate function depth; useful when combing through a
stack trace.
- effort to improve readability by inlining the `watchErrors` and
`shutdown` methods.

scull7 commented:
@ricejasonf 

- change most anonymous functions to be named functions.
- remove arbitrary function currying
- mark functions as private using the dangling ‘_’ character.
- multiple ‘_’ indicate function depth; useful when combing through a
stack trace.
- effort to improve readability by inlining the `watchErrors` and
`shutdown` methods.

ricejasonf commented on lib/pool.js:
```diff
@@ -0,0 +1,110 @@
+
+    return originalCreateMethod().catch((e) =>
+
+      new Bluebird(function ___backoffRunResolver(resolve, reject) {
```
Sorry to be a pest, but it looks like this `catch` should simply return `e` and chain to a promise instead of nesting the definition of a new promise inside the catch.

scull7 replied:
I use the new promise so that I have access to the `reject` function.  

```javascript
setTimeout(() => reject(e), wait)
```

I will see how it behaves, but I'm not sure that throwing an exception will result in the same context.

ricejasonf reviewed (approved):
:shipit:
